### PR TITLE
Nikos/5939/eip1193 provider callback types

### LINKED ISSUE #5939: [4.x] types Web3BaseProvider for EIP1193Provider has event wrong type in web3-types
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

https://github.com/web3/web3.js/blob/1e5a4c3c22d8d4b363395bf288f861e84f8acad1/packages/web3-types/src/web3_base_provider.ts#L158-L175

### Expected Behavior

to be alignment with the specification
https://eips.ethereum.org/EIPS/eip-1193#appendix-i-consumer-facing-api-documentation

### Steps to Reproduce

n/a

### Web3.js Version

4.x

### Environment

n/a

### Anything Else?

_No response_

### PR DESCRIPTION
**Pull request by nikoulai**
## Description
 fixes events' callback types for eip1193 providers
 closes #5939 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist for 1.x:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` with success.
- [ ] I have tested the built `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.

## Checklist for 4.x:

- [ ] I have selected the correct 4.x base branch.
- [ ] Within the description, the feature or issue is discussed in detail or an issue is linked.   
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added any required tests for the changes I made 
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `yarn` successfully
- [ ] I ran `yarn lint` successfully
- [ ] I ran `yarn build:web` successfully
- [ ] I ran `yarn test:unit` successfully
- [ ] I ran `yarn test:integration` successfully
- [ ] I ran `compile:contracts` successfully
- [ ] I have tested my code.
- [ ] I have updated the corresponding `CHANGELOG.md` file in the packages I have edited.

### CODE REVIEW

File: packages/web3-types/src/web3_base_provider.ts [hunk 1]
```diff
@@ -157,28 +160,32 @@ export abstract class Web3BaseProvider<API extends Web3APISpec = EthExecutionAPI
...
 	): void;
 	public abstract on(
 		type: 'connect' | 'chainChanged',
-		callback: Web3ProviderEventCallback<ProviderConnectInfo>,
+		callback: Web3Eip1193ProviderEventCallback<ProviderConnectInfo>,
 	): void;
 	public abstract on(
 		type: 'accountsChanged',
-		callback: Web3ProviderEventCallback<{
+		callback: Web3Eip1193ProviderEventCallback<{
```
Comment (BeroBurny) [inline packages/web3-types/src/web3_base_provider.ts]: seems that types are still wrong, if you check the [appendix of specification](https://eips.ethereum.org/EIPS/eip-1193#appendix-i-consumer-facing-api-documentation) you can observe different type

**Review by BeroBurny — CHANGES_REQUESTED** [review]

File: packages/web3-types/src/web3_base_provider.ts (hunk 1 above)
Comment (BeroBurny) [reply packages/web3-types/src/web3_base_provider.ts]: maybe the wallet connect example can help visualize the issue
https://docs.walletconnect.com/1.0/quick-start/dapps/web3-provider#events-eip-1193

File: packages/web3-types/src/web3_api_types.ts [hunk 2]
```diff
@@ -31,6 +31,7 @@ export interface EthSubscription extends ProviderMessage {
 }
 
 export interface ProviderRpcError extends Error {
+	message: string;
```
Comment (Muhammad-Altabba) [inline packages/web3-types/src/web3_api_types.ts]: The Error interface provided by TypeScript already contains a `message` (and also a `name`) 
```suggestion
```

Here is the definision of the Error interface:
```ts
interface Error {
    name: string;
    message: string;
    stack?: string;
}
```

File: packages/web3-types/src/web3_base_provider.ts [hunk 3]
```diff
@@ -159,27 +162,32 @@ export abstract class Web3BaseProvider<API extends Web3APISpec = EthExecutionAPI
...
+		listener: Web3Eip1193ProviderEventCallback<any> | Web3ProviderEventCallback,
 	): void;
-	public abstract removeListener(type: string, callback: Web3ProviderEventCallback): void;
 
-	public abstract once?<T = JsonRpcResult>(
+	public abstract once?<T>(
 		type: string,
-		callback: Web3ProviderEventCallback<T>,
+		callback: Web3Eip1193ProviderEventCallback<T> | Web3ProviderEventCallback<T>,
 	): void;
```
Comment (Muhammad-Altabba) [inline packages/web3-types/src/web3_base_provider.ts]: I think `removeListener` and `once` should have the same type restrictions as `on`.

File: packages/web3-utils/src/socket_provider.ts [hunk 4]
```diff
@@ -199,7 +200,10 @@ export abstract class SocketProvider<
 	 * @param type - The event type to listen for
 	 * @param callback - The callback to be invoked when the event is emitted
 	 */
-	public on<T = JsonRpcResult>(type: EventType, callback: Web3ProviderEventCallback<T>): void {
+	public on<T = JsonRpcResult>(
+		type: EventType,
+		callback: Web3Eip1193ProviderEventCallback<T> | Web3ProviderEventCallback<T>,
+	): void {
 		this._eventEmitter.on(type, callback);
 	}
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/src/socket_provider.ts]: I am thinking that having only one implementation for `on` will not force the eact type restrictions we have at `Ẁeb3BaseProvider`. What do you think about having here something like this?:

```suggestion
	public abstract on(
		type: 'disconnect',
		listener: Web3Eip1193ProviderEventCallback<ProviderRpcError>,
	): void {
		this._eventEmitter.on(type, listener);
	}
	public abstract on<T = JsonRpcResult>(
		type: 'message' | string,
		listener: Web3Eip1193ProviderEventCallback<ProviderMessage> | Web3ProviderEventCallback<T>,
	): void {
		this._eventEmitter.on(type, listener);
	}
	public abstract on(
		type: 'connect',
		listener: Web3Eip1193ProviderEventCallback<ProviderConnectInfo>,
	): void {
		this._eventEmitter.on(type, listener);
	}
	public abstract on(
		type: 'chainChanged',
		listener: Web3Eip1193ProviderEventCallback<string>,
	): void {
		this._eventEmitter.on(type, listener);
	}
	public abstract on(
		type: 'accountsChanged',
		listener: Web3Eip1193ProviderEventCallback<string[]>,
	): void {
		this._eventEmitter.on(type, listener);
	}
	public abstract removeListener(
		type: string,
		listener: Web3Eip1193ProviderEventCallback<any> | Web3ProviderEventCallback,
	): void {
		this._eventEmitter.on(type, listener);
	}
```

And also doing the same for `once` and `removeListener`?

**Review by avkos — APPROVED** [review]

File: packages/web3-utils/src/socket_provider.ts (hunk 4 above)
Comment (Muhammad-Altabba) [reply packages/web3-utils/src/socket_provider.ts]: It has been applied on `once` but not on `removeListener`. Is there a specific reason? Or was it just missed by a chance?

File: packages/web3-utils/src/web3_eip1193_provider.ts [hunk 5]
```diff
@@ -104,6 +103,7 @@ export abstract class Eip1193Provider<
 			});
 	}
 
+	// todo this must be ProvideRpcError with a message too
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/src/web3_eip1193_provider.ts:106]: Could you please create an issue for this?

File: packages/web3-core/src/web3_subscriptions.ts [hunk 6]
```diff
@@ -27,6 +27,7 @@ import {
 	Log,
 	JsonRpcNotification,
 	JsonRpcSubscriptionResult,
+	Web3ProviderEventCallback,
```
Comment (Muhammad-Altabba) [inline packages/web3-core/src/web3_subscriptions.ts]: I think this is not needed
```suggestion
```

File: packages/web3-core/src/web3_subscriptions.ts [hunk 7]
```diff
@@ -81,7 +82,10 @@ export abstract class Web3Subscription<
 			}
 		};
 
-		(this._requestManager.provider as Web3BaseProvider).on<Log>('message', messageListener);
+		(this._requestManager.provider as Web3BaseProvider).on<Log>(
+			'message',
+			messageListener as Web3ProviderEventCallback<Log>,
```
Comment (Muhammad-Altabba) [inline packages/web3-core/src/web3_subscriptions.ts]: I think the casting is not needed
```suggestion
		(this._requestManager.provider as Web3BaseProvider).on<Log>('message', messageListener);
```

**Review by Muhammad-Altabba — APPROVED** [review]
LGTM.

However, there is tiny stuff. If you can address them, before merging.

File: packages/web3-utils/src/web3_eip1193_provider.ts (hunk 5 above)
Comment (nikoulai) [reply packages/web3-utils/src/web3_eip1193_provider.ts:106]: https://github.com/web3/web3.js/issues/6003

File: packages/web3-utils/src/socket_provider.ts [hunk 8]
```diff
@@ -197,28 +199,88 @@ export abstract class SocketProvider<
...
+		type: 'accountsChanged',
+		listener: Web3Eip1193ProviderEventCallback<string[]>,
+	): void;
+	public removeListener<T = JsonRpcResult>(
+		type: 'message' | string,
+		listener: Web3Eip1193ProviderEventCallback<ProviderMessage> | Web3ProviderEventCallback<T>,
+	): void;
+	public removeListener<T = JsonRpcResult>(
+		type: string,
+		listener: Web3Eip1193ProviderEventCallback<any> | Web3ProviderEventCallback<T>,
```
Comment (jdevcs) [inline packages/web3-utils/src/socket_provider.ts]:
```suggestion
		listener: Web3Eip1193ProviderEventCallback<unknown> | Web3ProviderEventCallback<T>,
```
?

File: packages/web3-utils/src/socket_provider.ts [hunk 9]
```diff
@@ -197,28 +199,88 @@ export abstract class SocketProvider<
...
+	public removeListener(
+		type: 'chainChanged',
+		listener: Web3Eip1193ProviderEventCallback<string>,
+	): void;
+	public removeListener(
+		type: 'accountsChanged',
+		listener: Web3Eip1193ProviderEventCallback<string[]>,
+	): void;
+	public removeListener<T = JsonRpcResult>(
+		type: 'message' | string,
```
Comment (jdevcs) [inline packages/web3-utils/src/socket_provider.ts]: type: 'message' | string,
I think it should be only 'message', or is there reason of allowing it string as well?

File: packages/web3-utils/src/socket_provider.ts [hunk 10]
```diff
@@ -197,28 +199,88 @@ export abstract class SocketProvider<
...
+		type: 'accountsChanged',
+		listener: Web3Eip1193ProviderEventCallback<string[]>,
+	): void;
+	public once<T = JsonRpcResult>(
+		type: 'message' | string,
+		listener: Web3Eip1193ProviderEventCallback<ProviderMessage> | Web3ProviderEventCallback<T>,
+	): void;
+	public once<T = JsonRpcResult>(
+		type: string,
+		listener: Web3Eip1193ProviderEventCallback<any> | Web3ProviderEventCallback<T>,
```
Comment (jdevcs) [inline packages/web3-utils/src/socket_provider.ts]:
```suggestion
		listener: Web3Eip1193ProviderEventCallback<unknown> | Web3ProviderEventCallback<T>,
```
`any` should be avoided where possible in our lib.

Comment (jdevcs) [reply packages/web3-utils/src/socket_provider.ts]: + similar change proposed for other parts of code in this PR

**Review by jdevcs — COMMENTED** [review]
lgtm, except few changes required. ( also update change log ) Thanks @nikoulai

**Pull request merged**